src/storage/store/block_engine.rs

```diff
@@ -132,34 +200,180 @@ impl BlockEngine {
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<proto::HubEvent, MessageValidationError> {
+        match message.msg_type() {
+            MessageType::LendStorage => {
+                Ok(self.stores.storage_lend_store.merge(message, txn_batch)?)
+            }
+            _ => return Err(MessageValidationError::InvalidMessageType),
+        }
+    }
+
+    fn on_merge_message(storage_slot: &mut StorageSlot, merge_message_body: &MergeMessageBody) {
+        if let Some(added_message) = &merge_message_body.message {
+            match added_message.data.as_ref().unwrap().body.as_ref().unwrap() {
+                proto::message_data::Body::LendStorageBody(lend_storage_body) => {
+                    storage_slot.sub(&StorageSlot::from_storage_lend(&lend_storage_body));
+                }
+                _ => {}
+            }
+        }
+
+        for deleted_message in &merge_message_body.deleted_messages {
+            match deleted_message
+                .data
+                .as_ref()
+                .unwrap()
+                .body
+                .as_ref()
+                .unwrap()
+            {
+                proto::message_data::Body::LendStorageBody(lend_storage_body) => {
+                    storage_slot.merge(&StorageSlot::from_storage_lend(&lend_storage_body));
+                }
+                _ => {}
+            }
+        }
+    }
+
     pub(crate) fn replay_snapchain_txn(
         &mut self,
         trie_ctx: &merkle_trie::Context,
         snapchain_txn: &Transaction,
         txn_batch: &mut RocksDbTransactionBatch,
-    ) -> Result<Vec<u8>, BlockEngineError> {
+        timestamp: &FarcasterTime,
+        version: EngineVersion,
+    ) -> Result<(Vec<u8>, Vec<HubEvent>), BlockEngineError> {
+        let mut hub_events = vec![];
         for message in &snapchain_txn.system_messages {
             if let Some(ref onchain_event) = message.on_chain_event {
                 match self
                     .stores
                     .onchain_event_store
                     .merge_onchain_event(onchain_event.clone(), txn_batch)
                 {
-                    Ok(event) => self
-                        .trie
-                        .update_for_event(trie_ctx, &self.db, &event, txn_batch)?,
+                    Ok(event) => {
+                        self.trie
+                            .update_for_event(trie_ctx, &self.db, &event, txn_batch)?;
+                        hub_events.push(event);
+                    }
                     Err(err) => {
                         error!("Unable to merge onchain event: {:#?}", err.to_string())
                     }
                 }
             }
         }
 
+        let mut storage_slot = self
+            .storage_slot_for_transaction(snapchain_txn, false)
+            .unwrap();
+
+        for message in &snapchain_txn.user_messages {
+            match self.validate_user_message(message, &storage_slot, timestamp, version, txn_batch)
```

Comment:
We also need to account for the user hitting the limits on the storage lend message itself, right? i.e. there's no pruning of storage lend messages happening right now.

And when prunes happen, we need to account for the deleted lend messages and emit block events for those.

Practically, I'm not sure if we'll hit these limits, but we can't allow infinite storage lend messages.